platform/android/export/export.cpp

```diff
@@ -768,6 +768,81 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		}
 	}
 
+	void _write_tmp_manifest(const Ref<EditorExportPreset> &p_preset, bool p_give_internet, bool p_debug) {
+		String manifest_text =
+				"<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
+				"<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
+				"    xmlns:tools=\"http://schemas.android.com/tools\">\n";
+
+		manifest_text += _get_screen_sizes_tag(p_preset);
+		manifest_text += _get_gles_tag();
+
+		Vector<String> perms;
+		_get_permissions(p_preset, p_give_internet, perms);
+		for (int i = 0; i < perms.size(); i++) {
+			manifest_text += vformat("    <uses-permission android:name=\"%s\" />\n", perms.get(i));
+		}
+
+		bool uses_xr = (int)(p_preset->get("xr_features/xr_mode")) == 1;
+		if (uses_xr) {
+			int dof_index = p_preset->get("xr_features/degrees_of_freedom"); // 0: none, 1: 3dof and 6dof, 2: 6dof
+			if (dof_index == 1) {
+				manifest_text += "    <uses-feature tools:node=\"replace\" android:name=\"android.hardware.vr.headtracking\" android:required=\"false\" android:version=\"1\" />\n";
+			} else if (dof_index == 2) {
+				manifest_text += "    <uses-feature tools:node=\"replace\" android:name=\"android.hardware.vr.headtracking\" android:required=\"true\" android:version=\"1\" />\n";
+			}
+			int hand_tracking_index = p_preset->get("xr_features/hand_tracking"); // 0: none, 1: optional, 2: required
+			if (hand_tracking_index == 1) {
+				manifest_text += "    <uses-feature tools:node=\"replace\" android:name=\"oculus.software.handtracking\" android:required=\"false\" />\n";
+			} else if (hand_tracking_index == 2) {
+				manifest_text += "    <uses-feature tools:node=\"replace\" android:name=\"oculus.software.handtracking\" android:required=\"true\" />\n";
+			}
+		}
+
+		String package_name = p_preset->get("package/unique_name");
+		manifest_text += vformat(
```

Comment:
There are a lot of sections of the Android Manifest that can be refactored out of the `_write_tmp_manifest` method and into helper methods in `gradle_export_util.h`. For example, `_get_gles_tag()` and `_get_screen_sizes_tag()`. Is it better to keep that functionality inside this method, or split it up and move it outside of this method?

Keeping it inside this method is arguably easier to understand, but it adds clutter to an already very large file.

Reply:
I favor a modular approach so feel free to move tag generation into their own method like you did for `_get_gles_tag()` and `_get_screen_sizes_tag()`.

Larger tags like `<application>` and `<activity>` should have their own method that return the entire tag properly populated to this method. For example a `_get_application_tag()` method should internally call a `_get_activity_tag()` to include the `<activity>` tag.

A nice side effect is that it gives a quick high level overview of all the various components supported in the Godot `AndroidManifest` file.